Use Alpha and Beta for our prerelease tags instead of dev and preview for all languages

weshaggard opened the pull request:
preview in the name of packages can potentially cause confusion
because some preview azure service features generally have to
charge less then full price for the service. Since the SDK versions
aren't exactly tied to preview or GA versions of the service we want
to avoid this confusion and just use the word beta for our pre-GA
package versioning.

weshaggard commented:
@ramya-rao-a @bterlson @xirzec could I get someone from JS team to review and approve this?

xirzec commented:
I think the only outstanding issue is the fact that a dependency set to "^1.0.0-beta.1" will automatically roll forward to "1.0.0-dev.20200720.1" since it is semantically higher and compatible, which would be unexpected since dailies are generally not stable or well supported.

Internally we avoid this by pinning preview versions, but unless a user invokes npm with `--save-exact`, they will not get that behavior and might end up installing a daily instead of a preview release. 

Are you comfortable renaming `preview` to `alpha` in order to mitigate?

weshaggard commented:
> Are you comfortable renaming dev to alpha in order to mitigate?

I would like to avoid renaming dev to alpha, dev has a particular meaning that indicates it is a daily dev build and I'd like that to also remain consistent with the other languages. 

I will say that the problem you call out is an problem with the other languages as well but it is slightly mitigated there because the other languages require you to add an additional feed source in order to get dev builds. npm doesn't allow us to easily segregate and have packages from multiple feeds for the same scope so we had to publish the dev packages to npm directly under the dev tag. 

How much of an issue do you think it will be for folks that depend on "^1.0.0-beta.1" to roll-forward to the latest dev version? Aren't they already sort of signing up for any prerelease by using "^" on a prerelease, by its nature that is already accepting risk because we don't guarantee compat between prereleases. I understand our dev builds might be a little more volatile but I'm not sure if that is a big enough concern to not make this change. If customers do hit issues they can always pin to a beta version like we do.

xirzec commented:
> 
> How much of an issue do you think it will be for folks that depend on "^1.0.0-beta.1" to roll-forward to the latest dev version? Aren't they already sort of signing up for any prerelease by using "^" on a prerelease, by its nature that is already accepting risk because we don't guarantee compat between prereleases. I understand our dev builds might be a little more volatile but I'm not sure if that is a big enough concern to not make this change. If customers do hit issues they can always pin to a beta version like we do.

I'm not sure I am swayed by your argument. The idea of semver is to have semantically meaningful versions. We're abusing semver in this case to provide nightlies, because as you mention it's a pain to release nightlies otherwise and npm has no support for release channels.

While you are correct that we allow breaking changes between prereleases, we at least document these changes and provide upgrade advice in the changelog. This is not necessarily true for every nightly snapshot of master as sometimes the release notes are in flux until we ship a prerelease. Also prereleases (generally) occur at a frequency of once a month, rather than every day.

I'm not sure I fully understand the motivation for this change in the first place. Why would the version string used by an SDK package have anything to do with the billing status of an Azure service? We have never aligned to the version strings of APIs themselves.

Also, many consumer products have coopted the use of `beta` to mean pre-release and not fully functional. If a user might expect some discount for using "preview" software, why would they not demand the same of a beta?

xirzec commented:
One last thought - you use the term `prerelease` - a nice word that starts with a 'p'. Why don't we use that instead of `beta` or `preview` if the only trouble is the English word "preview"?

weshaggard commented:
Thanks for the feedback @xirzec. I know for at least python and java "beta" has special meaning in their versions as they don't fully support semver, which is why they were already not using "preview". Given that we were already using beta for those languages we chose to try and align on beta for the other 2 languages given the feedback about the extra meaning of "preview" from a service point of view.

@kurtzeborn @Petermarcu what are your guys thoughts on moving from "preview" to "prerelease"? It would avoid the issue that the storage team pointed out and not have the potential roll forward issue @xirzec points out for JS. 

@AlexGhiondea would .NET have any issue with using "prerelease" instead of "beta" or "preview"?

kurtzeborn commented:
Looks like "beta" may be a problem for JS.  I'm fine with the word "prerelease" instead of "preview" for .NET and JS then.  I'll ping @amishra-dev and @pemari-msft and see if they have concerns.

kurtzeborn commented:
We talked about this in the storage sync today and using the word "prerelease" for JS is fine with the storage team.  Their preference is to use "bata" for .NET since that matches Java (and Python somewhat).  They also think that using the word "beta" for other langauges (C/C++, Go, mobile) makes the most sense since it would match the rest (unless there are idiomatic reasons not to).

weshaggard commented:
Based on various feedback we decided to move all languages to `beta` except JS and keep them with `preview` as they are due to the dev versioning conflict called out by @xirzec.

weshaggard commented:
OK after may conversations on this top I've updated the guidelines based on the following:

To summarize the changes:
For preview releases we will use “-beta” as the prerelease tag for all languages. Instead of “-preview”. 
For daily dev builds we will use “-alpha” as the prerelease tag for all languages. Instead of “-dev”.

The move from “preview” to “beta” is to make the languages consistent and is to avoid confusion with using “preview” as it has meaning for Azure services that indicates people get discounted rates. The move from “dev” to “alpha” is to have the versions that sort older then “beta” and helps ensure that our daily dev versions, which are less stable, don’t end up as the “latest” version for our customers when they are testing out our previews. Folks will have to take an extra opt-in step if they wish to test our daily dev versions, it won’t happen by accident like it can today. 

I've also added a note in each language section on how folks can remain on the daily dev version train if they want to opt-in to those versions.

Can I please get the language leads and/or architects to review and sign-off on this PR documenting the changes?

- .NET @AlexGhiondea @KrzysztofCwalina 
- Java @Joshfree @JonathanGiles 
- JS/TS @ramya-rao-a @bterlson 
- Python @lmazuel @johanste

weshaggard commented:
FYI @Azure/azure-sdk-eng

ramya-rao-a commented on docs/policies/releases.md:
```diff
@@ -185,14 +183,14 @@ Below are the guidelines for versions and tags to use:
+  - If a package has moved from beta to GA, ensure that the `next` tag is deleted.
+- Beta releases will use the format `X.Y.Z-beta.N` for version and the published package should have the tag `next`.
+  - Additionally, use the `latest` tag **only** if the package has never had a GA release.
+- Daily dev releases will use the format `X.Y.Z-alpha.YYYYMMDD.r` (`r` is based on the number of builds performed on the given day) and the published package will have the tag `dev` and published to npm. To consume a dev package either pin to a specific version or use the `dev` tag as the version to avoid crossing packages with other beta packages. You can also use a version range similar to `X.Y ...
```
> use the `dev` tag as the version to avoid crossing packages with other beta packages

This is not clear to me. What are we trying to avoid here...?

> You can also use a version range similar to `X.Y.Z-alpha - X.Y.Z-beta` to get the latest dev release and avoid the beta release versions

What is a scenario where this would be helpful?

Generally, can we avoid the entire new text that is being added here? We can make the assumption that the user knows how to use semver and how to specify the dependency ranges in their package.json file.

weshaggard replied:
alpha/dev versions are not always earlier then the beta versions as they are two different versioning trains and so generally if you want to be on the alpha/dev version then you want to say on that train and not accidently start picking up beta versions. To do that you either need to use the `dev` tag or use the version range specifier. The reason I put the version range specifier directly here is because it is not a normal case folks would do and I wanted people to have an example based on our versioning guidelines.

xirzec replied:
I also don't think we need to expand here, as I don't believe folks tend to use ranges with pre-release versions. Especially since both alpha and beta packages of ours can have breaking changes, rolling forward automatically would always be a bad idea.

I'm also a little confused as to who this text is targeted at. Internally, we **always** pin pre-release dependencies to exact versions when we must use them.

ramya-rao-a replied:
`dev` tag information is fine, it does not add any ambiguity. I would remove the part on the range.

weshaggard commented on docs/policies/releases.md:
```diff
@@ -104,22 +102,22 @@ The release cycle of each SDK Component may vary based on the needs of the under
-- Dev: a build containing the most up-to-date changes based on the current master branch. Expect frequent and potentially breaking change in this release.
-- Preview: a release generated to get customer feedback before a GA. Preview releases are revised less often than dev. Preview releases may have breaking changes from the previous preview, but should not have breaking changes from the last GA release. Once a package has released to GA, any breaking changes require an exception and approval from the architecture board.
+- Beta: A release generated to get customer feedback before a GA. Beta releases are revised less often than dev. Beta releases may have breaking changes from the previous beta, but should not have breaking changes from the last GA release. Once a package has released to GA, any breaking changes require an exception and approval from the architecture board.
+- Dev: A release generated to give customers early access to our most up-to-date changes. Customers should not take production dependencies on these as they are volatile and will break often. They should only be used to temporarily unblock issues or to get a early read on the upcomming releases.
```
I'm torn about changing Dev to alpha everywhere as I still would like to make the distinction that the alpha/dev is a unique versioning train. Similar to why you requested we stick with the npm "dev" tag instead of switching it to "alpha" it helps convey the message that they are intended to be developer versions only for testing.

>The point on production dependencies due to volatility applies to both alpha and beta, but is listed only for the former

I mostly want to call out that alpha/dev versions are more volatile then the beta versions. I will add a note to that effect.

I can reorder them these or update the text so the context is a little clearer.

weshaggard commented on docs/policies/releases.md:
```diff
@@ -142,30 +140,30 @@ Each language repo uses a badge for analysis of dependencies on `master` includi
+Packages which depend on beta versions should pin specifically to the beta version on which they depend because beta releases may contain breaking changes.
 
-Avoid source dependencies in projects and only use binary/package dependencies. If a package under development needs to incorporate changes from a package upon which it depends then it should use the nightly dev version of that package. That is, if Project B incorporates changes to Project A then Project B should consume the _Package_ from Project A that is released in the nightly feed.
+Avoid source dependencies in projects and only use binary/package dependencies. If a package under development needs to incorporate changes from a package upon which it depends then it should use the dev version of that package. That is, if Project B incorporates changes to Project A then Project B should consume the _Package_ from Project A that is released in the dev feed.
```
I've been either removing the word nightly or changing it to daily depending on the context.  In this context I didn't feel like the fact it was done daily added any value.

weshaggard commented on docs/dotnet/introduction.md:
```diff
@@ -771,7 +771,7 @@ Consistent version number scheme allows consumers to determine what to expect fr
 {% include requirement/MUST id="dotnet-version-semver" %} use _MAJOR_._MINOR_._PATCH_ format for the version of the library dll and the NuGet package.
 
-Use _-preview_._N_ suffix for preview package versions. For example, _1.0.0-preview.2_.
+Use _-beta._N_ suffix for pre-release package versions. For example, _1.0.0-beta.2_.
```
I don't remember why I did that but you are correct I called this beta everywhere else so I will update this.

ramya-rao-a commented on docs/policies/releases.md:
```diff
@@ -104,22 +102,26 @@ The release cycle of each SDK Component may vary based on the needs of the under
-In addition to standard SemVer, the team occasionally releases a preview of a package to allow the community to dog food and give feedback on new features.
+In addition to the stable GA releases the team also has prereleases of a package to allow the community to try new features early and give feedback.
+
+- Alpha releases are sometimes referred to as dev releases and use a prerelease label that contains a date stamp similar to `-alpha-YYYYMMDD.r` to ensure the versions are unique as they will often be published daily. These are often published to an isolated registry depending on the language ecosystem. These releases are based on the latest committed code changes and should not be used for produc ...
```
Nit: I recommend breaking up sentences for an easier read
```suggestion
- Alpha releases are sometimes referred to as dev releases and use a prerelease label that contains a date stamp similar to `-alpha-YYYYMMDD.r`. This ensures the versions are unique as they are published daily. These are often published to an isolated registry depending on the language ecosystem. These releases are based on the latest committed code changes and should not be used for production dependencies. They are very volatile and change from version-to-version. These are mostly useful to temporarily working around an issue or testing out the latest library changes. 
```

ramya-rao-a commented on docs/policies/releases.md:
```diff
@@ -104,22 +102,26 @@ The release cycle of each SDK Component may vary based on the needs of the under
+In addition to the stable GA releases the team also has prereleases of a package to allow the community to try new features early and give feedback.
+
+- Alpha releases are sometimes referred to as dev releases and use a prerelease label that contains a date stamp similar to `-alpha-YYYYMMDD.r` to ensure the versions are unique as they will often be published daily. These are often published to an isolated registry depending on the language ecosystem. These releases are based on the latest committed code changes and should not be used for produc ...
+- Beta releases use a prerelease label like `-beta.X` and are published to the most common public registry for each language ecosystem. These releases are less volatile and released less often then alpha releases and are usually used before releasing a new minor or major GA release. Beta releases may have breaking changes from the previous beta but should not have breaking changes from the last G ...
```
Nit: Breaking up sentences for an easier read:

```suggestion
- Beta releases use a prerelease label like `-beta.X` and are published to the most common public registry for each language ecosystem. These releases are less volatile and released less often then alpha releases. These are usually used before releasing a new minor or major GA release. Beta releases may have breaking changes from the previous beta but should not have breaking changes from the last GA release. Once a package has released to GA, any breaking changes require an exception and approval from the architecture board.
```

ramya-rao-a commented on docs/policies/releases.md:
```diff
@@ -142,30 +144,30 @@ Each language repo uses a badge for analysis of dependencies on `master` includi
+Packages which depend on beta versions should pin specifically to the beta version on which they depend because beta releases may contain breaking changes.
 
-Avoid source dependencies in projects and only use binary/package dependencies. If a package under development needs to incorporate changes from a package upon which it depends then it should use the nightly dev version of that package. That is, if Project B incorporates changes to Project A then Project B should consume the _Package_ from Project A that is released in the nightly feed.
+Avoid source dependencies in projects and only use binary/package dependencies. If a package under development needs to incorporate changes from a package upon which it depends then it should use the alpha version of that package. That is, if Project B incorporates changes to Project A then Project B should consume the _Package_ from Project A that is released as alpha to the dev feed.
```
>  If a package under development needs to incorporate changes from a package upon which it depends then it should use the alpha version of that package. That is, if Project B incorporates changes to Project A then Project B should consume the _Package_ from Project A that is released as alpha to the dev feed.

Who is this advice for? Are both Project A & B packages that we are working on? In JS, I dont see a scenario where we need to use the alpha version of another package.

weshaggard replied:
I think this guidance is a little outdated. I will update it because I agree there are not any scenarios were we would have checked in alpha dependency versions in the repo. The main idea was to try and avoid having source level dependencies because it becomes challenging as you have to build everything and it becomes too easy to depend on new features accidently. This is part of the reason we added the min/max testing in all the languages to help notify us when we are in this situation. The primary goal is to make sure our dependency versions (or ranges) are correct and that we ship all the necessary dependencies in the correct order, especially if you depend on new features.

ramya-rao-a commented on eng/scripts/release-template/python.md:
```diff
@@ -16,13 +16,13 @@ The Azure SDK team is pleased to make available the %%MMMM yyyy%% client library
 ## Installation Instructions
 
-To install the latest preview version of the packages, copy and paste the following commands into a terminal:
+To install the latest beta version of the packages, copy and paste the following commands into a terminal:
```
Why does the python template's Installation Instructions focus on beta?

weshaggard replied:
Not sure that would be a question for the python team.

weshaggard commented:
Thanks for all the feedback. I'm going to override the CI failure and merge as the broken link is unrelated to these changes and I've fixed in another PR.
